### DISCUSSION TITLE
Languages

### DISCUSSION THREAD

**Discussion by Krowne** [iBotPeaches/Apktool#3216] [Ideas]
Would it be possible to add a file with all the texts that are used, to be able to create translations?

The idea would be to create the file with variables, and each variable contains a text. Each user who is interested in downloading and building the project could then translate this file into their language.

Also, as has been commented in another thread, the implementation of colors, depending on whether it is an information message, a warning, an error, etc. For example, the information "I" in blue, the warning "W" and the error "E" in red, etc... but the message texts are always white.

I made a program that translates everything, and adds the colors depending on the type of information displayed on the screen, the problem is that when you add changes to the texts or just add texts, I have to do a search again to find out what the changes are. If all the texts are in a single file as variables, it would be easier to locate them for translation.

If it's a lot of work and you don't want to mess around with it, I understand.

Thank you very much for the work you do.

**Comment by iBotPeaches** [comment]
You mean like all the strings in the application? I kinda lost you at the I/W/E stuff as not sure how that is related to strings and the bundling process.

Also not sure I follow how this would be implemented. Sure I can dump all strings much like aapt2/aapt can do in some form of a structured output like:

```
➜  Optimized aapt2 d strings app-debug.apk
String pool of 18 unique UTF-8 non-sorted strings, 18 entries and 0 styles using 824 bytes:
String #0 : res/drawable-v24/$ic_launcher_foreground__0.xml
String #1 : res/drawable/ic_launcher_background.xml
String #2 : res/drawable-v24/ic_launcher_foreground.xml
String #3 : res/mipmap-mdpi-v4/ic_launcher.webp
String #4 : res/mipmap-mdpi-v4/ic_launcher_round.webp
String #5 : res/mipmap-hdpi-v4/ic_launcher.webp
String #6 : res/mipmap-hdpi-v4/ic_launcher_round.webp
String #7 : res/mipmap-xhdpi-v4/ic_launcher.webp
String #8 : res/mipmap-xhdpi-v4/ic_launcher_round.webp
String #9 : res/mipmap-xxhdpi-v4/ic_launcher.webp
String #10 : res/mipmap-xxhdpi-v4/ic_launcher_round.webp
String #11 : res/mipmap-xxxhdpi-v4/ic_launcher.webp
String #12 : res/mipmap-xxxhdpi-v4/ic_launcher_round.webp
String #13 : res/mipmap-anydpi-v26/ic_launcher.xml
String #14 : res/mipmap-anydpi-v26/ic_launcher_round.xml
String #15 : ARSC Test
String #16 : res/xml/backup_rules.xml
String #17 : res/xml/data_extraction_rules.xml
```

However, the next question usually after that is "how do I build the application with my updated translations?". If that is the next question then why did we make an intermediate format to translate? You could easily just iterate the base `values/strings.xml` that would accomplish it.

**Comment by Krowne** [comment]
> You mean like all the strings in the application? I kinda lost you at the I/W/E stuff as not sure how that is related to strings and the bundling process.
> 
> Also not sure I follow how this would be implemented. Sure I can dump all strings much like aapt2/aapt can do in some form of a structured output like:
> 
> ```
> ➜  Optimized aapt2 d strings app-debug.apk
> String pool of 18 unique UTF-8 non-sorted strings, 18 entries and 0 styles using 824 bytes:
> String #0 : res/drawable-v24/$ic_launcher_foreground__0.xml
> String #1 : res/drawable/ic_launcher_background.xml
> String #2 : res/drawable-v24/ic_launcher_foreground.xml
> String #3 : res/mipmap-mdpi-v4/ic_launcher.webp
> String #4 : res/mipmap-mdpi-v4/ic_launcher_round.webp
> String #5 : res/mipmap-hdpi-v4/ic_launcher.webp
> String #6 : res/mipmap-hdpi-v4/ic_launcher_round.webp
> String #7 : res/mipmap-xhdpi-v4/ic_launcher.webp
> String #8 : res/mipmap-xhdpi-v4/ic_launcher_round.webp
> String #9 : res/mipmap-xxhdpi-v4/ic_launcher.webp
> String #10 : res/mipmap-xxhdpi-v4/ic_launcher_round.webp
> String #11 : res/mipmap-xxxhdpi-v4/ic_launcher.webp
> String #12 : res/mipmap-xxxhdpi-v4/ic_launcher_round.webp
> String #13 : res/mipmap-anydpi-v26/ic_launcher.xml
> String #14 : res/mipmap-anydpi-v26/ic_launcher_round.xml
> String #15 : ARSC Test
> String #16 : res/xml/backup_rules.xml
> String #17 : res/xml/data_extraction_rules.xml
> ```
> 
> However, the next question usually after that is "how do I build the application with my updated translations?". If that is the next question then why did we make an intermediate format to translate? You could easily just iterate the base `values/strings.xml` that would accomplish it.

If you can implement the strings in the resource path, it would be perfect, you should only update the list in English, in the strings.xml file, if any other user wants to create their language, they would only have to create the corresponding directory, copy the original strings.xml file to said directory and translate it.

In my case I would have to create the values-es directory and put the strings.xml file inside and then translate it from the one in the values directory.

Regarding I, W, E, I am going to try to add a screenshot, with the texts already converted to Spanish as I told you before, so that you understand what I was referring to with the colors.

https://i.imgur.com/m0LgT71.jpg

As I told you in the other thread of the color request, if you want, I could provide you with the color codes to implement for versions of Windows 10 onwards. The only drawback is that you have to add a key to the Windows registry for the system console to detect the color codes. But for this, the user could be informed of the command to add the activation of the colors in the console to the registry.

reg add "HKEY_CURRENT_USER\Console" /v VirtualTerminalLevel /t REG_DWORD /d 1 /f

**Reply by Krowne** [reply]
Also if you want, I can provide you with the last project that you uploaded, with the changes made in the code to show I/W/E according to its corresponding color, and also the way in which I made the translations to my language.

**Reply by iBotPeaches** [reply]
Okay so I understand the color portion of the stream formatter, but still lost a bit on the language stuff. The process you described with taking `values/strings.xml` and placing it into `values-xx/strings.xml` and translating seems like the best process.

**Reply by Krowne** [reply]
You must take into account that depending on the operating system console, the color codes must be implemented differently.

And yes, I think the most practical way to create multiple languages is by adding values-xx directories as you mentioned, and simply copying the strings.xml file and modifying the texts.
